zebra-state/src/service/memory_state.rs

```diff
@@ -460,11 +402,199 @@ impl Ord for Chain {
     }
 }
 
+/// The state of the chains in memory, incuding queued blocks.
+#[derive(Default)]
+pub struct NonFinalizedState {
+    /// Verified, non-finalized chains.
+    chain_set: BTreeSet<Box<Chain>>,
+}
+
+impl NonFinalizedState {
+    /// Finalize the lowest height block in the non-finalized portion of the best
+    /// chain and update all side-chains to match.
+    pub fn finalize(&mut self) -> Arc<Block> {
+        let chains = mem::take(&mut self.chain_set);
+        let mut chains = chains.into_iter();
+
+        // extract best chain
+        let mut best_chain = chains.next().expect("there's at least one chain");
```

Comment:
This iterator is in ascending order, that is, the chain with the lowest work is first:
https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.iter

But the best chain has the highest work.

Is there a test for the best chain?
If there is, that test is also incorrect.
If not, please add one.

It might also be helpful to create a separate `best_chain` function.

```suggestion
        let mut best_chain = chains.next_back().expect("there's at least one chain");
```

Reply:
Un-resolved, because this bug should also have a working test.

Reply:
added todo in tests follow up PR